linux/just_git_airgap_repo.bsh

```diff
@@ -0,0 +1,1507 @@
+#
+# .. code-block:: bash
+#
+#   export JUST_SETUP_SCRIPT="$(basename "${BASH_SOURCE[0]}")"
+#   if [ ! -f "$(dirname "${BASH_SOURCE[0]}")/external/vsi_common/env.bsh" ]; then
+#     echo "'just' could not be loaded. Trying to setup the repository as an"
+#     echo "air-gapped repository"
+#     # source the contents of repo_map.env (in a bash 3.2 compatible way)
+#     source /dev/stdin <<< \
+#         "$(git show origin/__just_git_mirror_info_file:repo_map.env 2>/dev/null || :)"
+#     if ! declare -Fx git_airgap_submodule_update; then
+#       echo "ERROR the vsi_common submodule could not be found!"
+#       return 1
+#     fi
+#     git_airgap_submodule_update external/vsi_common
+#   fi
+#   source "$(dirname "${BASH_SOURCE[0]}")/external/vsi_common/env.bsh"
+#
+# At this point, it is possible to run ``just git clone-airgap-repos`` to
+# recursively initialize and clone the submodules from their air-gapped mirrors
+#
+# REVIEW This could become a git hook; however, the hook would still have to be
+# added in the setup.env file
+#**
+
+function print_git_airgap_submodule_update()
+{
+  # NOTE Do not use vsi_common functionality in this function! It is saved to
+  # repo_map.env and sourced in setup.env
+  uwecho '#**
+          # .. function:: git_airgap_submodule_update
+          #
+          # Reconfigure and update a submodule
+          #
+          # :Arguments: ``$1`` - A relative path (from the top-level repo) to a submodule
+          # :Parameters: ``repo_paths`` - The (relative) path from the top-level repository
+          #                          to each submodule (recursively)
+          #              ``repo_urls`` - The respective url for each submodule (recursively)
+          #**
+          function git_airgap_submodule_update()
+          (
+            set -eu # fail fast
+
+            # Strip trailing slash
+            local submodule_path="${1%/}"
+
+            # Find submodule in the list
+            local repo_url
+            if declare -p repos &>/dev/null; then
+              repo_url="${repos["${submodule_path}"]}"
+              unset repos
+            else
+              local index
+              for index in "${!repo_paths[@]}"; do
+                if [ "${repo_paths[$index]}" = "${submodule_path}" ]; then
+                  repo_url="${repo_urls[$index]}"
+                  break
+                fi
+              done
+              unset repo_urls repo_paths
+            fi
+            if [ -z "${repo_url:+set}" ]; then
+              echo "ERROR submodule_path, ${submodule_path}, not found in repo_paths:"
+              for index in "${!repo_paths[@]}"; do echo "  ${repo_paths[$index]}"; done
+              return 1
+            fi
+
+            # This follows just:git_mirror:git_clone_main
+            git submodule init "${submodule_path}"
+            git config "submodule.${submodule_path}.url" "${repo_url}"
+            git submodule update "${submodule_path}"
+          )
+          '
```

Comment:
If you want, you can put this in it's own file in vsi_common, so it'll be a normal file, and if you want to, you can unit test it like normal, etc...

And then _just_ `cat` the file.

(As it stands, I'm actually surprised this doesn't break my bash autodocs.)

Reply:
Would that still require an function like this? If so, my point is the same, you can make is a separate file and cat that file instead of echoing like this.

Reply:
It might be better actually to orphan commit `git_mirror:update_submodules` (removing any dependent functionality) and the logic of the `just_git_airgap_repo:git_clone-airgap-repo` target (move it into a function), and then run that in the setup.env file. This would support the case when vsi_common is a sub-submodule